rholang/src/test/scala/coop/rchain/rholang/interpreter/accounting/CostAccountingSpec.scala

```diff
@@ -52,11 +52,23 @@ class CostAccountingSpec extends FlatSpec with Matchers with PropertyChecks with
 
   val contracts = Table(
     ("contract", "expectedTotalCost"),
-    ("""@0!(2)""", 33L),
-    ("""@0!(2) | @1!(1)""", 69L),
-    ("""for(x <- @0){ Nil }""", 64L),
-    ("""for(x <- @0){ Nil } | @0!(2)""", 73L),
-    /*
+    ("""@0!(2)""", 97L),
+    ("""@0!(2) | @1!(1)""", 197L),
+    ("""for(x <- @0){ Nil }""", 128L),
+    ("""for(x <- @0){ Nil } | @0!(2)""", 329L),
+    ("@0!!(0) | for (_ <- @0) { 0 }", 342L),
+    ("@0!!(0) | for (x <- @0) { 0 }", 342L),
+    ("@0!!(0) | for (@0 <- @0) { 0 }", 336L),
+    ("@0!!(0) | @1!!(1) | for (_ <- @0;_ <- @1) { 0 }", 594L),
+    ("@0!(0) | for (_ <- @0) { 0 }", 333L),
+    ("@0!(0) | for (x <- @0) { 0 }", 333L),
+    ("@0!(0) | for (@0 <- @0) { 0 }", 327L),
+    ("@0!(0) | for (_ <= @0) { 0 }", 354L),
+    ("@0!(0) | for (x <= @0) { 0 }", 356L),
+    ("@0!(0) | for (@0 <= @0) { 0 }", 341L),
+    ("@0!(0) | @0!(0) | for (_ <= @0) { 0 }", 574L),
+    ("@0!(0) | for (@0 <- @0) { 0 } | @0!(0) | for (_ <- @0) { 0 }", 663L),
+    ("@0!(0) | for (@0 <- @0) { 0 } | @0!(0) | for (@1 <- @0) { 0 }", 551L),
```

Comment:
I would add a few peek tests just to make sure those are also deterministic

Reply:
Check out the cost log for `("@0!(0) | for (@0 <<- @0) { 0 }", 328L)`:
```
List(Cost(30,parsing), Cost(11,send eval), Cost(11,receive eval), Cost(4,substitution), Cost(4,substitution), Cost(4,substitution), Cost(4,substitution), Cost(8,produces storage), Cost(4,substitution), Cost(14,consume storage), Cost(64,event storage cost), Cost(-14,consume storage refund), Cost(-8,produces storage refund), Cost(64,event storage cost), Cost(128,comm event storage cost))
```

The produce should not have been refunded for because peek isn't going to remove it from the tuplespace

Reply:
Great catch! :+1:

Reply:
What happens with multiple persistent sends or mixed sends?  E.g. `@0!!(0) | @0!!(0) | for (_ <- @0) { 0 }` or `@0!!(0) | @0!(0) | for (_ <- @0) { 0 }`.

Are you going to refund storage fees for stuff that can be garbage collected?

Reply:
Have any major changes been made to the cost accounting in the spatial matcher since https://github.com/rchain/rchain/pull/1501 ?  If so, it's worth testing that cost accounting there is also deterministic.

Reply:
I'll add the cases you proposed and make sure they're deterministic.

Garbage collection is out of scope for Mercury, so refunds based on it(s possibility) should be too, I think.

We're no longer charging for matching since the merge of #2637, as pointed out in one of the commit messages.

Reply:
`@0!!(0) | @0!!(0) | for (_ <- @0) { 0 }` passes, I've added it to the test suite

`@0!!(0) | @0!(0) | for (_ <- @0) { 0 }` can have different costs due to the execution result itself being non-deterministic. We need a different test for that, one that uses replay and an event log. I'd like to have it part of the next PR.

Reply:
@pyrocto please see the above :)

@lukasz-golebiewski I attempted fixing this, the fix is blocked by the bug I mentioned in https://discordapp.com/channels/391435113415573504/406508474261045250/615482010747666443. Let's pull the peek tests into the next PR, or the one coming after the bug fix.

Reply:
I'm not sure what you mean by "`!P` replication". Having the same process twice?

If yes, notice the `@0!!(0)`-s below are different processes
`@0!!(0) | @0!!(0) | for (_ <- @0) { 0 }`, `@0!!(0)` is always different than `@0!(0)`, and in all cases the `!0` events are different b/c they come from different namespaces. I feel I still don't get your point though? :)